### PR TITLE
Added jenkinsfile to run gradle check in OpenSearch

### PR DESCRIPTION
**Pull request by owaiskazi19**
Signed-off-by: Owais Kazi <owaiskazi19@gmail.com>

### Description
Created a new ubuntu docker image for gradle check. [Link](https://hub.docker.com/layers/opensearchstaging/ci-runner/ci-runner-ubuntu1804-build-v1/images/sha256-2c7bb2780bc08cd4e7e3c382ac53db414754dabd52f9b70e1c7e344dfb9a0e5e?context=explore)
Added a jenkinsfile to run gradle check in OpenSearch. After jenkins go public logs will be available in the CI.

Sample run: http://opens-jenki-a7sro6sneahz-2051798054.us-west-2.elb.amazonaws.com/job/Gradle_Check/job/Gradle_Check/4/console

Steps to run the jenkinsfile:
1. Create a pipeline job with Git as SCM. Add the opensearch repo url to clone, branch as `main`.
2. Provide the script path as `jenkins/jenkinsfile`.

### Issues Resolved
Closes https://github.com/opensearch-project/OpenSearch/issues/2507

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: jenkins/gradle-check-assemble.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
```
Comment (dblock) [inline jenkins/gradle-check-assemble.sh:1]: Should this file be called `gradle-check.sh`?

File: jenkins/gradle-check-assemble.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+IFS=`echo -ne "\n\b"`
+
+docker ps
+docker stop `docker ps -qa` > /dev/null 2>&1
```
Comment (dblock) [inline jenkins/gradle-check-assemble.sh:5]: Why is this needed?

**Review by dblock — APPROVED** [review]

File: jenkins/gradle-check-assemble.sh (hunk 1 above)
Comment (setiah) [inline jenkins/gradle-check-assemble.sh:1]:
```suggestion
#!/usr/bin/env bash
```
#!/usr/bin/env searches PATH for bash, and bash is not always in /bin, particularly on non-Linux systems. For example, on my OpenBSD system, it's in /usr/local/bin, since it was installed as an optional package.

https://stackoverflow.com/questions/21612980/why-is-usr-bin-env-bash-superior-to-bin-bash/21613044#21613044

File: jenkins/post-build-task2.sh [hunk 3]
```diff
@@ -0,0 +1,60 @@
+#!/bin/bash
```
Comment (setiah) [inline jenkins/post-build-task2.sh:1]: same as previous comment

File: jenkins/gradle-check-assemble.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+cd search
```
Comment (saratvemulapalli) [inline jenkins/gradle-check-assemble.sh:7]: Similar question, why is this needed?

File: jenkins/gradle-check-assemble.sh [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+echo "./gradlew check --no-daemon --no-scan"
+./gradlew check --no-daemon --no-scan
+GRADLE_CHECK_STATUS=$?
+
+
+# Archive reports
+cd ../
+REPORTS_DIR=`find ./search -name reports`
+zip -9 -q -r gradle_check_${BUILD_NUMBER}_reports.zip ${REPORTS_DIR}
+ls | grep "gradle_check_${BUILD_NUMBER}"
```
Comment (saratvemulapalli) [inline jenkins/gradle-check-assemble.sh:18]: Not a blocker for this PR, just a thought:
Looks like we can publish code coverage results from here. 
cc: @tlfeng

File: jenkins/gradle-check-assemble.sh (hunk 4 above)
Comment (peterzhuamazon) [reply jenkins/gradle-check-assemble.sh:7]: I have stated many times that search is just a folder name of the repo.
If you just git clone opensearch then the default folder is opensearch not search.
The reason we use search back in the day because opensearch naming is not defined.

If you clone as opensearch (which seems like you do in the docker file), you should remove cd into search and cd into the corresponding folder instead.

File: jenkins/gradle-check-assemble.sh (hunk 1 above)
Comment (peterzhuamazon) [reply jenkins/gradle-check-assemble.sh:1]: I agree, this should not be gradle check assemble.

**Review by peterzhuamazon — CHANGES_REQUESTED** [review]
I see a lot of code directly copy paste from previous Jenkins without considering the circumstances of the new Jenkins and why the old code was being written that way.

I also would like to see some results of this workflow in practice before approving it.

Seems like the jenkins pipeline file is also written in `scripted` way instead of `decalarative` which makes the maintenance a bit harder.

Block this PR for now as I am concerned about the outcome of this PR.

File: jenkins/post-build-task2.sh [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+set -e
+
+OLDIFS=$IFS
+ISSUE_TITLE="Jenkins Randomized Gradle Check Failed on ${sha1}"
+
+###############################################
+
+function create_issue() {
```
Comment (peterzhuamazon) [inline jenkins/post-build-task2.sh:9]: This particular script was written in this way to make sure auto / random checks is running daily without adding a new workflow.

Also, this script is used based on the fact that ghprb is in action.
Have you considered how to setup ghprb on new Jenkins instances to reproduce the current setup?

And probably we should rewrite this as the old code is not in good shape but a temp fix solution.

Thanks.

File: jenkins/gradle-check-assemble.sh (hunk 4 above)
Comment (saratvemulapalli) [reply jenkins/gradle-check-assemble.sh:7]: Sure that makes sense, the question was to @owaiskazi19, do we still need this ?

**Comment by peterzhuamazon** [comment]
Hi Owais,

Also realize this is not even a Jenkins Pipeline file, but instead a dsl groovy which we are not using anymore.
1. Please follow the build repo `jenkins` folder to see examples of how to setup proper jenkinsfile
2. The workflow needs to be run within a docker container. Since you are running multiple containers within the runner container, you need to use the ubuntu buildx image we provided for docker build, see `jenkins/docker` for instances.
3. You should simplify the workflow and remove old scripts from the old workflow if they can be re-written or replaced in the new setup.
4. You pipeline should be able to run on our dev jenkins and should be testable.

Thanks.

cc: @bbarani

**Comment by owaiskazi19** [comment]
> 
Thanks for the input here @peterzhuamazon. I have pushed the changes.

File: jenkins/gradle-check-assemble.sh (hunk 4 above)
Comment (owaiskazi19) [reply jenkins/gradle-check-assemble.sh]: Changed the design. Created a new ubuntu docker image for gradle check purpose and run it in a jenkisnfile. We don't need the scripts anymore.

**Comment by owaiskazi19** [comment]
https://github.com/opensearch-project/OpenSearch/issues/1703

File: jenkins/jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,28 @@
+pipeline {
+    agent {
+         docker {
+            label 'AL2-X64'
+            image 'opensearchstaging/ci-runner:ci-runner-ubuntu1804-build-v1'
```
Comment (gaiksaya) [inline jenkins/jenkinsfile:5]: Maybe add link to this docker file: https://github.com/opensearch-project/opensearch-build/blob/main/docker/ci/dockerfiles/build.ubuntu18.opensearch.x64.dockerfile if anyone wants to see whats in the docker image

**Review by gaiksaya — APPROVED** [review]
Great to see such a short JenkinsFile :)

File: jenkins/jenkinsfile (hunk 7 above)
Comment (owaiskazi19) [reply jenkins/jenkinsfile:9]: Sounds good. Pushed the change. Thanks for the review :)

**Review by CEHENKLE — APPROVED** [review]

File: jenkins/jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,32 @@
+pipeline {
+    agent {
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply jenkins/jenkinsfile:2]: @abhinavGupta16 can you create an issue for the above so that it can be pick up later? Thanks

[the remark this answers is not in the document]
Comment (abhinavGupta16) [reply jenkins/jenkinsfile:2]: https://github.com/opensearch-project/OpenSearch/issues/2690
